rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md

```diff
@@ -1 +1,157 @@
-In review, see <https://github.com/nervosnetwork/rfcs/pull/238>, <https://github.com/nervosnetwork/rfcs/pull/236>, and <https://github.com/nervosnetwork/rfcs/pull/237>
+---
+Number: "0032"
+Category: Consensus (Hard Fork)
+Status: Draft
+Author: Ian Yang
+Organization: Nervos Foundation
+Created: 2021-04-26
+---
+
+# CKB VM Version Selection
+
+## Abstract
+
+This RFC proposes a mechanism to choose the CKB VM version after the hard fork. What's included in the new CKB VM version is beyond the scope of this document.
+
+## Motivation
+
+It's essential to keep improving CKB VM because it is the computation bottleneck of the whole network. The upgrade packages can improve the performance, bring bug fixings and add new RISC-V extensions. However the upgrade should not break the old code, users must have the opt-in option to specify the VM version.
+
+This RFC proposes a general mechanism that determines how the CKB node chooses the CKB VM version for a transaction script group.
+
+## Specification
+
+The CKB VM upgrades via hard fork. The next scheduled hard fork is ckb2021, which is activated since a specific epoch. For all the transactions in the blocks before the activation epoch, they must run the CKB VM version 0 to verify all the script groups. In these transactions, the `hash_type` in cell lock and type script must be either 0 or 1 in the serialized molecule data.
```

Comment:
There is a mention of `CKB VM version 0` out of nowhere, I do recommend we put up a short section at the start of this specification, mentioning that the `Lina` fork version uses CKB VM version 0, while the new ckb2021 fork introduces a new CKB VM version 1.

Reply:
Also it's worth mentioning that the selection of CKB VM versions, is enabled via `hash_type`. Otherwise one can get confused why this paragraph mentions `hash_type` at all, which seems to be completely irrelavant to CKB VM versions.